13_sandboxes/sandbox_pool.py

```diff
@@ -0,0 +1,231 @@
+# a sandbox from the pool, getting a URL to the server running in the sandbox.
+#
+# Maintaining a pool of warm sandboxes is useful for example if your sandboxes need
+# to do significant work after being created, like downloading code, installing
+# dependencies, or running tests, before they are ready to serve requests.
+#
+# It uses a [Modal Queue](https://modal.com/docs/guide/dicts-and-queues#modal-queues)
+# to store references to the warm sandboxes, and functionality to maintain the pool
+# by adding and removing sandboxes, checking the current size, etc.
+#
+# The pool keeps track of the time to live for each sandbox, and will always return
+# a sandbox with at least 5 minutes left.
+#
+# ## Setting things up
+#
+# Start by deploying the Modal app:
+#
+# ```bash
+# modal deploy 13_sandboxes/sandbox_pool.py
+# ```
+#
+# This deploys the app with an empty pool. To fill the pool with 3 sandboxes, run
+#
+# ```bash
+# modal run 13_sandboxes/sandbox_pool.py::resize_pool --target 3
+# ```
+#
+# You can check the current size of the pool by running:
+#
+# ```bash
+# modal run 13_sandboxes/sandbox_pool.py::check_pool --verbose
+# ```
+#
+# ## Claiming a sandbox form the pool
+#
+# You can claim a sandbox by sending a GET request to the web endpoint URL.
+
+
+import time
+from datetime import datetime
+
+import modal
+
+app = modal.App("sandbox-pool")
+
+pool_queue = modal.Queue.from_name("sandbox-pool-buffer", create_if_missing=True)
+
+server_image = modal.Image.debian_slim(python_version="3.11").pip_install(
+    "fastapi[standard]~=0.115.14",
+    "requests~=2.32.4",
+)
+sandbox_image = modal.Image.debian_slim(python_version="3.11")
+
+
+SERVER_PORT = 8080
+HEALTH_CHECK_TIMEOUT_SECONDS = 10
+SANDBOX_TIMEOUT_SECONDS = 5 * 60
+SANDBOX_USE_DURATION_SECONDS = 2 * 60
+
+
+# ## Health check
+#
+# In this example, we run a very simple health check that just ensures that the
+# server is running and responding to requests.
+def health_check_sandbox(daemon_url):
+    import requests
+
+    start_time = time.time()
+    while time.time() - start_time < HEALTH_CHECK_TIMEOUT_SECONDS:
+        try:
+            response = requests.get(daemon_url, timeout=HEALTH_CHECK_TIMEOUT_SECONDS)
+            response.raise_for_status()
+            return
+        except requests.RequestException:
+            if time.time() - start_time >= HEALTH_CHECK_TIMEOUT_SECONDS:
+                raise
+            time.sleep(0.1)
+
+    raise requests.RequestException("Health check failed")
+
+
+# ## Adding a sandbox to the pool
+#
+# This function is called by the `resize_pool` function to add a sandbox to the pool,
+# and also by the `get_sandbox` function to replace the sandbox that was claimed.
+#
+# It creates a new sandbox, runs the health check, and adds the sandbox to the pool.
+@app.function(image=server_image, retries=3)
+@modal.concurrent(max_inputs=100)
+def add_sandbox_to_queue():
+    # This is done so we don't create sandboxes in the ephemeral app
```

Comment:
Nice note. We need to be better about documenting this kind of thing. Why use this as opposed to the `app` which is already in-context though? My assumption was that you would deploy this, so this app would live as long as the Sandboxes in the pool would.